main.cpp

```diff
@@ -210,285 +239,379 @@ struct Museum
     float impedance = 8.f;
-    // 5) the cabinets maximum reproduceable frequency (float)
     float maxFrequency = 6000.f;
 
-    // 1) convert voltage and current to sound
     double convertPowerToSound( float inputVoltage );
-    // 2) adjust load attenuator
     void setAttenuatorValue( float value );
-    // 3) release magic smoke
     void releaseMagicSmoke();
+private:
+    bool _speakersNotBlown = true;
 };
-/*
-Thing 10) Tube Guitar Amp Combo
-5 properties:
-    1) Power Supply
-    2) Output Section
-    3) Preamp Section
-    4) EQ Controls (or Tone Stack)
-    5) Speaker Cabinet
-3 things it can do:
-    1) Turn on / off
-    2) Adjust guitar EQ
-    3) Amplify guitar sound
- */
+
+double SpeakerCabinet::convertPowerToSound(float inputVoltage)
+{
+    if(_speakersNotBlown)
+    {
+        float wattageRequsted = inputVoltage * (inputVoltage / impedance);
+        double pascalsPerWatt = 2; // - using pascals because dB logarithmic stuff hard without framework to do math for me
+        if(wattageRequsted <= maxWattage * 1.1f)
+        {
+            // convert votage to assuming 2 pascals per watt 
+            return(static_cast<double>(inputVoltage) * (pascalsPerWatt * static_cast<double>(wattageRequsted)));
+        }
+        else
+        {
+            releaseMagicSmoke();
+            return 0;
+        }
+    }
+    return 0;
+}
+
+void SpeakerCabinet::releaseMagicSmoke()
+{
+    _speakersNotBlown = false;
+}
+
 struct TubeGuitarAmpCombo
 {
-    // 1) Power Supply
     PowerSupply ps;
-    // 2) Output Section
     OutputSection outputSection;
-    // 3) Preamp Section
     PreampSection preampSection;
-    // 4) EQ Controls (or Tone Stack)
     EQControls eq;
-    // 5) Speaker Cabinet
     SpeakerCabinet cabinet;
 
-    // 1) Turn on / off
     void setPowerState( bool value );
-    // 2) Adjust guitar EQ
     void adjustEQ( int band, float value );
-    //3) Amplify guitar sound
     double amplifyGuitarSound( float guitarSignal );
 };
+
+void TubeGuitarAmpCombo::setPowerState(bool value)
+{
+    ps.setPowerState(value);
+    if(ps.heaterOutputVoltage > 0.f)
+    {
+        outputSection.warmUpTubes(200.f);
+        preampSection.warmUpTubes(100.f);
+    }
+}
+
+void TubeGuitarAmpCombo::adjustEQ(int band, float value)
+{
+    if (band == 0) // lowSetting
+    {
+        eq.adjustBassSetting(value);
+    }
+    else if (band == 2) // trebleSetting
+    {
+        eq.adjustTrebleSetting(value);
+    }
+}
+
+double TubeGuitarAmpCombo::amplifyGuitarSound(float guitarSignal)
+{
+    if (ps.fuseState && ps.heaterOutputVoltage > 0.f && ps.mainOutputVoltage > 0.f)
+    {
+        float preampOutput = preampSection.amplifyGuitarSignalToLineLevel(guitarSignal);
+        float powerAmpOutput = outputSection.amplifyLineLevelAudioToSpeakerLevel(preampOutput);
+        return cabinet.convertPowerToSound(powerAmpOutput);
+    }
+    else // amp is off or blown
+    {
+        return 0;
+    }
+}
```

Comment:
If you find yourself writing code like this:
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    else 
    {
        return anotherValue;
    }
}
```
You do not need the `else` block.  Here's why:
`return` statements are the last thing that **ever happens** in a function.
When the `if()` expression is evaluated as `true` and the block is entered, the `return aValue;` is the last line of code to be executed before the function exits.
but when the `if()` expression evaluates as `false` and the block is **NOT** entered, the next thing that happens is `return anotherValue;`
you have no code after the `else` block that needs executing, since your `else` block contains a `return` statement.
that means you do **NOT** need the `else` block, and it is just causing visual clutter.
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    
    return anotherValue;
}
```
- [ ] revise